initialization/initialization_test.go

```diff
@@ -41,425 +46,466 @@ func (l testLogger) Debug(msg string, args ...interface{}) { l.t.Logf("\tDEBUG\t
+		WithLogger(testLogger{t: t}),
+	)
 	r.NoError(err)
-	init.SetLogger(log)
-
-	eg, _ := errgroup.WithContext(context.Background())
-
-	// call SessionNumLabelsWrittenChan() during initialization
-	eg.Go(func() error {
-		ch := make([]<-chan uint64, 0, 10)
-
-		for i := 0; i < 10; i++ {
-			time.Sleep(600 * time.Millisecond)
-			ch = append(ch, init.SessionNumLabelsWrittenChan())
-		}
-
-		if ch[0] != nil {
-			return errors.New("channel should be nil")
-		}
-
-		for _, c := range ch[1:] {
-			<-c
-		}
-		return nil
-	})
-	eg.Go(func() error {
-		time.Sleep(1 * time.Second)
-		return init.Initialize()
-	})
-	r.NoError(eg.Wait())
-
-	// call SessionNumLabelsWrittenChan() after initialization
-	eg.Go(func() error {
-		ch := make([]<-chan uint64, 0, 10)
-
-		for i := 0; i < 10; i++ {
-			ch = append(ch, init.SessionNumLabelsWrittenChan())
-		}
 
-		for _, c := range ch {
-			select {
-			case <-c:
-			default:
-				return errors.New("channel should be closed")
+	// Start initialization and stop it after it has written some labels
+	{
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		var eg errgroup.Group
+		eg.Go(func() error {
+			timer := time.NewTicker(50 * time.Millisecond)
+			defer timer.Stop()
+
+			for {
+				select {
+				case <-ctx.Done():
+					return nil
+				case <-timer.C:
+					if init.SessionNumLabelsWritten() > 0 {
+						cancel()
+						return nil
+					}
+				}
 			}
-		}
-		return nil
-	})
-	r.NoError(eg.Wait())
+		})
+		r.ErrorIs(init.Initialize(ctx), context.Canceled)
+		eg.Wait()
+
+		status := init.Status()
+		assert.Equal(t, StatusStarted, status)
+	}
 
-	// Cleanup.
-	r.NoError(init.Reset())
+	// Continue the initialization to completion.
+	{
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		var eg errgroup.Group
+		eg.Go(assertNumLabelsWritten(ctx, t, init))
+		r.NoError(init.Initialize(ctx))
+		cancel()
+		eg.Wait()
+
+		status := init.Status()
+		assert.Equal(t, StatusCompleted, status)
+	}
 }
 
-func assertNumLabelsWrittenChan(init *Initializer, t *testing.T) func() error {
+func assertNumLabelsWritten(ctx context.Context, t *testing.T, init *Initializer) func() error {
 	return func() error {
-		// hack to avoid receiving a nil channel from SessionNumLabelsWrittenChan()
-		// TODO (mafa): for a proper fix see https://github.com/spacemeshos/post/issues/78
-		var labelsChan <-chan uint64
-		assert.Eventually(t, func() bool {
-			labelsChan = init.SessionNumLabelsWrittenChan()
-			return labelsChan != nil
-		}, time.Second, time.Millisecond)
-
+		timer := time.NewTimer(50 * time.Millisecond)
+		defer timer.Stop()
 		var prev uint64
-		for p := range labelsChan {
-			assert.Less(t, prev, p)
-			prev = p
-			t.Logf("num labels written: %v\n", p)
+
+		for {
+			select {
+			case <-ctx.Done():
+				return nil
+			case <-timer.C:
+				num := init.SessionNumLabelsWritten()
+				t.Logf("num labels written: %v\n", num)
+				assert.LessOrEqual(t, prev, num)
```

Comment:
The intent is to check if `num` is increasing over time. IMO, this check should be inverted to match the intent :) something like:
```suggestion
				assert.GreaterOrEqual(t, num, prev)
```